File: apis/musics-service/src/channels/grpc/Grpc.ts
import * as grpc from 'grpc';

import IGrpcChannel from './interface';
import { MusicsHandler, MusicsService } from './handlers/MusicsHandler';
import Config from '@config/index';
import IErrorHandler from '@handlers/ErrorHandler/interface';
import ILoggerProvider from '@providers/LoggerProvider/interface';
import IAlbumsService from '@services/AlbumsService/interface';
import IArtistsService from '@services/ArtistsService/interface';
import IMusicsService from '@services/MusicsService/interface';

export default class Grpc implements IGrpcChannel {
  private server: grpc.Server;

  private errorHandler: IErrorHandler;
  private loggerProvider: ILoggerProvider;
  private musicsHandler: MusicsHandler;

  // prettier-ignore
  constructor(
    albumsService: IAlbumsService,
    artistsService: IArtistsService,
    musicsService: IMusicsService,
    errorHandler: IErrorHandler,
    loggerProvider: ILoggerProvider,
  ) {
    this.musicsHandler = new MusicsHandler(albumsService, artistsService, musicsService, errorHandler, loggerProvider);

    this.errorHandler = errorHandler;
    this.loggerProvider = loggerProvider;

    this.server = new grpc.Server();
  }

  public start(): void {
    this.server.addService(MusicsService, this.musicsHandler);

    const PORT = Config.channels.grpc.port;
    const HOST = Config.channels.grpc.host;

    // prettier-ignore
    this.server.bindAsync(
      `${HOST}:${PORT}`,
      grpc.ServerCredentials.createInsecure(),
      (error: Error | null, port: number): void => {
        if (error != null) {
          this.errorHandler.handleError(error);
        } else {
          this.loggerProvider.info(`gRPC server is running on port ${port}.`);
        }
      },
    );

    this.server.start();
  }

  public async stop(): Promise<void> {
    this.loggerProvider.info('Stopping gRPC server ...');

    return new Promise((resolve, reject) => {
      this.server.tryShutdown(() => {
        this.loggerProvider.info('gRPC server stopped.');
        resolve();
      });
    });
  }
}

File: apis/playlists-service/internal/channels/grpc/grpc.go
package grpc

import (
	"fmt"
	"net"

	"google.golang.org/grpc"
	"google.golang.org/grpc/reflection"
	"google.golang.org/grpc/status"

	"playlists-service/internal/channels/grpc/proto"
	"playlists-service/internal/config"
	"playlists-service/internal/constants"
	e "playlists-service/internal/handlers/errorHandler"
	l "playlists-service/internal/providers/loggerProvider"
	s "playlists-service/internal/services/playlistsService"
)

type grpcChannel struct {
	server           *grpc.Server
	errorHandler     e.IErrorHandler
	loggerProvider   l.ILoggerProvider
	playlistsService s.IPlaylistsService
}

func New(errorHandler e.IErrorHandler, loggerProvider l.ILoggerProvider, playlistsService s.IPlaylistsService) IGrpcChannel {
	return grpcChannel{
		errorHandler:     errorHandler,
		loggerProvider:   loggerProvider,
		playlistsService: playlistsService,
	}
}

func (c grpcChannel) Start() {
	listener, err := net.Listen("tcp", fmt.Sprintf(":%s", config.GrpcPort))
	if err != nil {
		panic(err)
	}

	c.server = grpc.NewServer()
	proto.RegisterPlaylistsServer(c.server, c)
	reflection.Register(c.server)

	if err := c.server.Serve(listener); err != nil {
		panic(err)
	}

	c.loggerProvider.Info(fmt.Sprintf("gRPC Channel running on port %s", config.GrpcPort))
}

func (c grpcChannel) Stop() {
	c.server.Stop()
}

func (c grpcChannel) handleError(err error) error {
	c.errorHandler.HandleError(err)

	var customError *constants.BaseError

	if !c.errorHandler.IsTrustedError(err) {
		customError = constants.InternalError
	} else {
		var ok bool
		customError, ok = err.(*constants.BaseError)
		if !ok {
			customError = constants.InternalError
		}
	}

	return status.Error(translateGrpcError(customError.StatusCode), customError.Message)
}
